### DISCUSSION TITLE
How far are we with Opus in-stream metadata?

### DISCUSSION THREAD

**Discussion by Moonbase59** [savonet/liquidsoap#3990] [Q&A]
I noticed @toots was active in https://trac.ffmpeg.org/ticket/10363 (and others) 14 months ago.

Opus has a superior sound quality and rather low bandwidth, but I can’t find any way to include metadata using Liquidsoap/AzuraCast, which makes it rather unusable. Players should be able to display metadata, and the [Opus spec indicates they can do in-stream metadata](https://opus-codec.org/docs/opusfile_api-0.7/structOpusTags.html#details).

Is there any "trick" I might have missed? My current Opus output looks like this:

```
output.icecast(%opus(samplerate=48000, bitrate=128, vbr="constrained", application="audio", channels=2, signal="music", complexity=10, max_bandwidth="full_band"), id="local_6", host = "127.0.0.1", port = 8000, password = "(PASSWORD)", mount = "/radio.opus", name = "Nite Radio", description = "Wir spielen die bessere Musik.", genre = "Classic Rock, Easy Listening, Electronic, Folk Rock, Hard Rock, Oldies, Pop, Schlager, Various", url = "https://grav1.example.com/", public = false, encoding = "UTF-8", send_icy_metadata=true, radio)
```
(Password and domain changed.)

Would be really fine after many years to finally get Opus streams with metadata. Apparently others can do: 

```
 ffprobe -hide_banner http://mustafejen.se:8000/dull.opus
Input #0, ogg, from 'http://mustafejen.se:8000/dull.opus':
  Metadata:
    icy-description : Hobby stream
    icy-genre       : Alternative, eclectic
    icy-name        : dull opus
    icy-pub         : 1
    icy-url         : https://mustafejen.se/~per
  Duration: N/A, start: 148.480000, bitrate: N/A
  Stream #0:0: Audio: opus, 48000 Hz, stereo, fltp
    Metadata:
      artist          : Noijjk
      title           : The Cosmic Microwave Background Radiation'
```

**Comment by vitoyucepi** [comment]
Hi @Moonbase59,
As far as I know, opus is a combination of ogg+opus, and ogg has the ability to transfer metadata internally. I think `curl http://mustafejen.se:8000/dull.opus | head -c16000 | less` is sufficient to get a quick look at the internals of an ogg stream.

**Comment by smoralis** [comment]
My observation is there a 2 types of opus streams.
Below my test streams
1. Chained opus streams  in ogg container which can have complete metadata inserted with Icecast Server
```
$ ffprobe -hide_banner http://zeremy.serv00.net:8200/opus
Input #0, ogg, from 'http://zeremy.serv00.net:8200/opus':
  Metadata:
    icy-br          : 128
    icy-description : Experimental Streaming Testing
    icy-genre       : Experimental Radio
    icy-name        : LSFB Experimental Radio [OPUS CHAINED 128]
    icy-pub         : 1
    icy-url         : https://zeremy.serv00.net
  Duration: N/A, start: 459.000000, bitrate: N/A
  Stream #0:0: Audio: opus, 48000 Hz, stereo, fltp
      Metadata:
        album           : Images and Words
        url             : http://coverartarchive.org/release/4b5a4d0e-1268-4ed5-8b48-6d0740053813/4163627164-500.jpg
        year            : 1992
        artist          : Dream Theater
        title           : Take the Time
        cover_url       : http://coverartarchive.org/release/4b5a4d0e-1268-4ed5-8b48-6d0740053813/4163627164-500.jpg
        track           : 03
        encoder         : Lavf60.3.100
``` 
2. Continuous ogg opus stream updating StreamTitle with artist-title with RSAS server
```
$ ffprobe -hide_banner http://zeremy.serv00.net:8202/opus
Input #0, ogg, from 'http://zeremy.serv00.net:8202/opus':
  Metadata:
    icy-br          : 128
    icy-description : Experimental Streaming Testing
    icy-genre       : Experimental Radio
    icy-name        : LSFB Experimental Radio [OPUS ICY 128]
    icy-pub         : 1
    icy-samplerate  : 44100
    icy-url         : https://zeremy.serv00.net
    StreamTitle     : Dream Theater - The Count of Tuscany
  Duration: N/A, start: 1308.000000, bitrate: N/A
  Stream #0:0: Audio: opus, 48000 Hz, stereo, fltp
      Metadata:
        encoder         : Lavf60.3.100
```

**Comment by smoralis** [comment]
Regarding Azuracast, I believe the default is
settings.encoder.metadata.export.set(["artist","title","album","song"])

my liquidsoap script uses
settings.encoder.metadata.export := ["artist", "title", "album", "genre", "date", "tracknumber", "comment", "track", "year", "dj", "next", "apic", "pic", "metadata_url", "metadata_block_picture", "coverart", "cover_url" , "url"]

**Comment by smoralis** [comment]
Also Azuracast uses IcecastKH, and doesn't accept icy url updates for ogg streams. 
You need to set send_icy_metadata=false.

**Reply by vitoyucepi** [reply]
> Also Azuracast uses IcecastKH, and doesn't accept icy url updates for ogg streams. You need to set send_icy_metadata=false.

Are you sure about this?

**Reply by smoralis** [reply]
Yes,  updating  via e.g
http://localhost:8888/admin/metadata?mount=/opus&mode=updinfo&song=Metallica%20-%20Chasing%20Light will return the url doesn't accept url updates and the metadata isn't accepted. 
With ogg opus or ogg flac.

**Comment by Moonbase59** [comment]
Thanks for the insights, @smoralis – real helpful.

So I set
```
output.icecast(%opus(samplerate=48000, bitrate=128, vbr="constrained", application="audio", channels=2, signal="music", complexity=10, max_bandwidth="full_band"), id="local_6", host = "127.0.0.1", port = 8000, password = "(PASSWORD)", mount = "/radio.opus", name = "Nite Radio", description = "Wir spielen die bessere Musik.", genre = "Classic Rock, Easy Listening, Electronic, Folk Rock, Hard Rock, Oldies, Pop, Schlager, Various", url = "https://grav1.example.com/", public = false, encoding = "UTF-8", send_icy_metadata=false, radio)
```

and indeed metadata is shown:

```
$ ffprobe -hide_banner https://radio.niteradio.net/listen/niteradio/radio.opus
Input #0, ogg, from 'https://radio.niteradio.net/listen/niteradio/radio.opus':
  Metadata:
    icy-description : Wir spielen die bessere Musik.
    icy-genre       : Classic Rock, Easy Listening, Electronic, Folk Rock, Hard Rock, Oldies, Pop, Schlager, Various
    icy-name        : Nite Radio
    icy-pub         : 0
    icy-url         : https://grav1.niteradio.net/
  Duration: N/A, start: 13.820000, bitrate: N/A
  Stream #0:0: Audio: opus, 48000 Hz, stereo, fltp
    Metadata:
      album           : Weird
      artist          : Penny Reign
      title           : Intangible
```

It seems that the bug hasn’t been fixed yet, if I leave it running the metadata starts to get cumulated:

```
$ mpv https://radio.niteradio.net/listen/niteradio/radio.opus
 (+) Audio --aid=1 'The Rescue' (opus 2ch 48000Hz)
File tags:
 Artist: Eureka
 Album: Shackleton's Voyage
 Title: The Rescue
AO: [pulse] 48000Hz stereo 2ch float
A: 00:02:38 / 00:02:40 (99%) Cache: 1.9s/80KB
[lavf] Linearizing discontinuity: -0.080000 -> 192.860000
A: 00:02:40 / 00:02:43 (98%) Cache: 2.6s/105KB
File tags:
 Artist: Eureka;Poems for Laila
 Album: Shackleton's Voyage;La Fillette triste
 Title: The Rescue;Sorrow
A: 00:05:29 / 00:05:31 (99%) Cache: 2.0s/83KB
[lavf] Linearizing discontinuity: 192.860000 -> 364.000000
A: 00:05:31 / 00:05:34 (99%) Cache: 2.5s/104KB
File tags:
 Artist: Eureka;Poems for Laila;Al Stewart
 Album: Shackleton's Voyage;La Fillette triste;Between the Wars
 Title: The Rescue;Sorrow;Sampan
A: 00:05:45 / 00:05:48 (99%) Cache: 2.5s/104KB
```

and some players like Audacious display the metadata from when the connection started, then never update again.

VLC somehow seems to do it right, changes metadata when it changes on stream. No wonder probably, because they use their own decoders.

**Anything that can be done? Or wait another few years?**

@smoralis You’re using Icecast 2.4.0-kh19, how do you get the `StreamTitle` into it (for ICY mode)? Maybe that could be the (worse) temporarily alternative.

**Reply by smoralis** [reply]
Regarding chained streams:
With VLC you might hear a small click sound when the next Metadata is received.
You can see the error messages in the player debug log.
Other players as you mentioned will append the data. 
foobar2000 and AIMP work. 

Using RSAS server all players accept the metadata. 
The RSAS server injects icy streamtitle. 
You can use liquidsoap with drop_metadata(s) and send the metadata to the server 
https://rocketbroadcaster.com/streaming-audio-server/docs/api.html#metadata-update

**Comment by smoralis** [comment]
http://zeremy.serv00.net:8200/opus
Is the Icecast 2.4.0-kh19, it is chained there's no streamtitle - not possible. 

http://zeremy.serv00.net:8202/opus
Is the RSAS server , with streamtitle.

**Reply by Moonbase59** [reply]
Ah ok, thanks again!

**Reply by Moonbase59** [reply]
@smoralis Mind keeping the "chained Opus" (port 8200) up for a while? Seems Chromium/Chrome can play that fine, but Firefox might want to triage. Appears that they have lost the capability starting with Firefox 57. I added to the 5- and 7-year-old bug reports [here](https://bugzilla.mozilla.org/show_bug.cgi?id=1427527#c9) and [here](https://bugzilla.mozilla.org/show_bug.cgi?id=1561291#c13).

I have not too much hope, since they start behaving a little like ffmpeg… "unloved" bugs are ignored for 5, 7, 12 _years_…

Anyway, we need Opus for its advantages, and we need metadata.

Thanks again for the insights, and for providing the test streams!

**Reply by smoralis** [reply]
@Moonbase59
No problem, streams back online!
Put in preview mode, playing 30 second samples so you have frequent metadata changes.

**Reply by Moonbase59** [reply]
You’re real nice—much obliged, thanks! (I also ran 2 test streams for a Firefox 115 bug for around 6 months, and that actually got fixed.)

**Comment by Moonbase59** [comment]
Let’s cross our fingers. You might also _involve_ on the bugs mentioned, and add a "confirm" or something, so they see it is actually something people _want_.

**Comment by smoralis** [comment]
@Moonbase59
Could you verify that my stream stops also with your Firefox version ? 
I can reproduce what you observed with your stream but mine doesn't.
Using Firefox for Android 127.0.1 (Build #2016027519), hg-b8d39a885c81+
GV: 127.0.1-20240618110440
AS: 127.0

**Reply by Moonbase59** [reply]
Hm, now that’s interesting.
- Your stream plays through
- My stream "stops" after first song (FF still reading stream but no sound)

Used Firefox 172.0.2 20240624183754 mint-001

I’m using Liquidsoap 2.2.5 and Icecast 2.4.0-kh22 (20240213100508) and this setting in LS:
```
output.icecast(%opus(samplerate=48000, bitrate=128, vbr="constrained", application="audio", channels=2, signal="music", complexity=10, max_bandwidth="full_band"), id="local_6", host = "127.0.0.1", port = 8000, password = "(PASSWORD)", mount = "/radio.opus", name = "Nite Radio", description = "Wir spielen die bessere Musik.", genre = "Classic Rock, Easy Listening, Electronic, Folk Rock, Hard Rock, Oldies, Pop, Schlager, Various", url = "https://grav1.niteradio.net/", public = false, encoding = "UTF-8", send_icy_metadata=false, radio)
```

I also only have _artist_, _title_ and _album_ in the metadata (default).

If it helps, these are the headers I see from my stream link:
```
HTTP/2 200 
server: openresty
date: Wed, 26 Jun 2024 12:53:33 GMT
content-type: application/ogg
pragma: no-cache
expires: Thu, 19 Nov 1981 08:52:00 GMT
cache-control: no-store, no-cache, private
vary: Origin
icy-description: Wir spielen die bessere Musik.
icy-genre: Classic Rock, Easy Listening, Electronic, Folk Rock, Hard Rock, Oldies, Pop, Schlager, Various
icy-name: Nite Radio
icy-pub: 0
icy-url: https://grav1.niteradio.net/
x-xss-protection: 1
x-content-type-options: nosniff
referrer-policy: no-referrer-when-downgrade
x-clacks-overhead: GNU Terry Pratchett, Leonard Nimoy
```

and these the headers I see on your stream:
```
HTTP/1.0 200 OK
Server: Icecast
Connection: Close
Pragma: no-cache
Expires: Thu, 19 Nov 1981 08:52:00 GMT
Cache-Control: no-store, no-cache, private
Vary: Origin
Access-Control-Allow-Origin: *
Date: Wed, 26 Jun 2024 12:50:28 GMT
Content-Type: application/ogg
icy-br: 128
ice-audio-info: channels=2;samplerate=48000;bitrate=128
icy-description: Experimental Streaming Testing
icy-genre: Experimental Radio
icy-name: LSFB Experimental Radio [OPUS CHAINED 128]
icy-pub: 1
icy-url: https://zeremy.serv00.net
```

Just tested Firefox/Android 127.0.1 (Build #20160227519), GV: 127.0.1-20240618110440 with same results.

Chromium, Chrome, Opera, Dolphin and even MS Edge also play my stream continuously.
The Tor Browser (actually Firefox 115-ESR) doesn’t.

**Reply by Moonbase59** [reply]
@smoralis Are you using Liquidsoap to stream? Maybe we could compare our `output.icecast()` lines? Really wonder why your stream now works.

**Reply by smoralis** [reply]
Yes, Using Liquidsoap version 2.2.5 in Windows 

and

icecast-kh-icecast-2.4.0-kh20
compiled from source 

cat /etc/os-release
NAME=FreeBSD
VERSION="13.2-RELEASE-p10"
VERSION_ID="13.2"
ID=freebsd

```
log.stdout.set(true)
log.file.set(true)

settings.encoder.metadata.export := ["artist", "title", "album", "genre", "date", "tracknumber", "comment", "track", "year", "dj", "next", "apic", "pic", "metadata_url", "metadata_block_picture", "coverart", "cover_url" , "url"]

settings.harbor.bind_addrs := ["0.0.0.0"]

s = mksafe(buffer(input.portaudio(device_id=3)))

s = insert_metadata(s)

def set_metadata(request)
log.important("Inserting metadata: #{request.query}")
s.insert_metadata(request.query)
http.response(data="Done.")
end

harbor.http.register.simple(port = 8090, "/set_metadata", set_metadata)


#Icecast

genre = "Experimental Radio"
description = "Experimental Streaming Testing"
host = "zeremy.serv00.net"
port = 8200
url = "https://zeremy.serv00.net"
password = 

#aac
 output.icecast(%ffmpeg(
  format="adts",
  %audio(samplerate = 44100, codec = "aac" , b = "320k")),
  host = host, 
  port = port,
  fallible=false,
  name= "LSFB Experimental Radio [AAC 320]",
  description = description,
  genre = genre,
  url = url,
  password = password,
  mount= "aac",
  send_icy_metadata=true,
  icy_metadata=["song", "title", "artist", "genre", "url"],
  s)

#mp3  
output.icecast(%ffmpeg(
  format="mp3",
  %audio(samplerate = 44100, codec = "libmp3lame", b = "320k")),  
  host = host, 
  port = port,
  fallible=false,
  name= "LSFB Experimental Radio [MP3 320]",
  description = description,
  genre = genre,
  url = url,
  password = password,
  mount= "mp3",
  send_icy_metadata=true,
  icy_metadata=["song", "title", "artist", "genre", "url"],
  s)
  
#opus
output.icecast(%ffmpeg(
  format="ogg",
  %audio(samplerate = 48000, codec = "libopus", b = "128k")),
  host = host, 
  port = port,
  fallible=false,
  name= "LSFB Experimental Radio [OPUS CHAINED 128]",
  description = description,
  genre = genre,
  url = url,
  password = password,
  mount= "opus",
  send_icy_metadata=false,
  s)
  
  #opus RSAS
output.icecast(%ffmpeg(
  format="ogg",
  %audio(samplerate = 48000, codec = "libopus", b = "128k")),
  host = host, 
  port = 8202,
  fallible=false,
  name= "LSFB Experimental Radio [OPUS ICY 128]",
  description = description,
  genre = genre,
  url = url,
  password = password,
  mount= "opus",
  send_icy_metadata=false,
  source.drop.metadata(s))
  


 

```

**Reply by Moonbase59** [reply]
Thanks for that! So you use ffmpeg and libopus instead of the %opus encoder I use. Have to play with that…

**Comment by Moonbase59** [comment]
@toots @vitoyucepi Could you tell us which encoder is actually used when using `%opus()`? Looks like that one has some problems (maybe using ffmpeg’s internal "opus" implementation?).

If I recode from
```ruby
enc = %opus(
  samplerate=48000,
  bitrate=128,
  vbr="constrained",
  application="audio",
  channels=2,
  signal="music",
  complexity=10,
  max_bandwidth="full_band"
)
output.icecast(
  enc,
  radio,
  id="local_7",
  host = "127.0.0.1",
  port = 8000,
  password = "(PASSWORD)",
  mount = "/test.opus",
  name = "Nite Radio",
  description = "Wir spielen die bessere Musik.",
  genre = "Classic Rock, Easy Listening, Electronic, Folk Rock, Hard Rock, Oldies, Pop, Schlager, Various",
  url = "https://grav1.niteradio.net/",
  public = false,
  encoding = "UTF-8",
  send_icy_metadata=false
)
```

to using `libopus` (v1.3.1 in my case, although they’re at 1.5.1 now):
```ruby
enc = %ffmpeg(format="ogg", %audio(
  codec="libopus",
  samplerate=48000,
  b="128k",
  vbr="constrained",
  application="audio",
  channels=2,
  compression_level=10,
  cutoff=20000
))
output.icecast(
  enc,
  radio,
  id="local_7",
  host = "127.0.0.1",
  port = 8000,
  password = "(PASSWORD)",
  mount = "/test.opus",
  name = "Nite Radio",
  description = "Wir spielen die bessere Musik.",
  genre = "Classic Rock, Easy Listening, Electronic, Folk Rock, Hard Rock, Oldies, Pop, Schlager, Various",
  url = "https://grav1.niteradio.net/",
  public = false,
  encoding = "UTF-8",
  send_icy_metadata=false
)
```

even Firefox has a smooth playout without stalling at metadata changes! The same thing @smoralis found with his setup.

@smoralis (and others) could you please verify that

- Firefox _doesn’t_ work with https://radio.niteradio.net/listen/niteradio/radio.opus
- Firefox _does_ work with https://radio.niteradio.net/listen/niteradio/test.opus

In the meantime, I’ll check out some more browsers.

**Reply by gAlleb** [reply]
On MacOS 1st url stops in Firefox. 2nd - doesn't. In Chrome - both work well.

**Reply by smoralis** [reply]
1st url stops.
2nd - doesn't.
Firefox for Android

**Reply by vitoyucepi** [reply]
> actually used when using `%opus()`

- `%opus` uses 
  1. `ocaml-opus`
  1. `ocaml-ogg`
  1. `libogg`
  1. `libopus`
- `%ffmpeg(format="ogg", %audio(codec="libopus"))` uses 
  1. `ocaml-ffmpeg`
  1. `ocaml-av`-something
  1. `libav`-something
  1. `libogg`
  1. `libopus`

**Reply by Moonbase59** [reply]
Thanks @vitoyucepi for the pointers! If they’re both using libopus, I wonder where the difference comes from. Maybe something in `ocaml-opus`…

**Comment by Moonbase59** [comment]
Tested so far and work well:
- Tor Browser (Firefox 115-ESR)*
- Firefox 115 (Linux, Windows)*
- Firefox 127.0.2 Desktop (Linux, MacOS, Windows)*
- Firefox Mobile 127 (misc. Android versions)*
- Chromium 125, 126 Desktop
- Chrome 124, 125, 126 Desktop (Windows, MacOS)
- Chrome Mobile 121 (Android, iOS)
- Chrome Webview 55.0, Android
- Opera 111 Desktop (Linux)
- Dolphin Android
- MS Edge 125 (Win10)
- MS Edge 126 (Win11)
- Logitech Media Server (now Lyrion Music Server) 8.5.2

All good on both streams, *=stalls on `radio.opus` stream but plays `test.opus` stream.

**So it seems we should use the `%ffmpeg` `libopus` encoder instead of `%opus` until we know whats happening there.**

**Comment by Moonbase59** [comment]
Here are some easy ways to check for stream and metadata.

Check metadata, using the official Opus decoder:

```
$ curl -s https://radio.niteradio.net/listen/niteradio/test.opus | opusdec --float - - >/dev/null
```

Check metadata and play in WAV format:

```
$ curl -s https://radio.niteradio.net/listen/niteradio/test.opus | opusdec --float --force-wav - - | ffplay -v quiet -
```

(You could substitute something else instead of `ffplay`.)

**Comment by smoralis** [comment]
FYI
Regarding FLAC in ogg chained streams the following patch for Icecast was reported 

https://gitlab.xiph.org/xiph/icecast-server/-/issues/2493

I tested the patch and it fixes the problem discussed 
about if we could also use Flac with the same method as Opus. 
Also the below thread is closed, if you could notify about the possibility :
https://github.com/AzuraCast/AzuraCast/issues/7240#issuecomment-2198200768

**Comment by mustafejen** [comment]
I noticed one of my streams was mentioned here. I am rusty when it comes to programming, but plenty of my code can be dug out from the web page https://mustafejen.se by right clicking in Firefox and choosing the right option in the meny which pops up.

The icecast configration file is not posted, but if I remember things right there was a trick in my icecast.xml.

The trick was
defining a mount point section for the opus mount point
and
adding type audio/ogg and subtype opus to the mount point section

     <mount type="normal">
    <mount-name>/dull.opus</mount-name>
    <fallback-mount>/tull.opus</fallback-mount>
    <fallback-override>1</fallback-override>
    <public>1</public>
    <stream-name>dull opus</stream-name>
    <stream-description>Hobby stream</stream-description>
    <stream-url>https://mustafejen.se</stream-url>
    <genre>Alternative, eclectic</genre>
    <bitrate>vbr</bitrate>
    <type>audio/ogg</type>
    <subtype>opus</subtype>
    <hidden>0</hidden>
    <burst-size>65536</burst-size>
    </mount>

This is how my mount section for dull.opus looks. You may have to modify it to get things working.

My apologies for necroposting
